main/src/observables/gravitational_waves.hpp

```diff
@@ -0,0 +1,195 @@
+{
+    T* x        = d.x.data();
+    T* y        = d.y.data();
+    T* z        = d.z.data();
+    T* vx       = d.vx.data();
+    T* vy       = d.vy.data();
+    T* vz       = d.vz.data();
+    T* gradP_x  = d.grad_P_x.data();
+    T* gradP_y  = d.grad_P_y.data();
+    T* gradP_z  = d.grad_P_z.data();
+    T* m        = d.m.data();
+
+    T out = 0.0;
+
+    std::array<T*, 3> pos = {x, y, z};
+    std::array<T*, 3> vel = {vx, vy, vz};
+    std::array<T*, 3> gradP = {gradP_x, gradP_y, gradP_z};
+
+    if(pos1 == pos2)
+    {
+
+#pragma omp parallel for reduction(+ : out)
+        for(size_t i = begin; i < end; i++)
+        {
+            T acc1 = -gradP[pos1][i];
+            T factor1 = vx[i] * vx[i] + vy[i] * vy[i] + vz[i] * vz[i];
+            T factor2 = x[i] * (-gradP_x[i]) + y[i] * (-gradP_y[i]) + z[i] * (-gradP_z[i]);
+
+            out += (3.0 * (vel[pos1][i] * vel[pos1][i] + pos[pos1][i] * acc1) - factor1 - factor2) * m[i];
+
+        }
+        return out * 2.0 / 3.0;
+    }
+    else
+    {
+#pragma omp parallel for reduction(+ : out)
+        for(size_t i = begin; i < end; i++)
+        {
+            T acc1 = -gradP[pos1][i];
+            T acc2 = -gradP[pos2][i];
+            out += (2.0 * vel[pos1][i] * vel[pos2][i] + acc1 * pos[pos2][i] + pos[pos1][i] * acc2) * m[i];
+        }
+        return out;
+    }
+}
+
+/*!@brief calculates the gravitational waves observable
+ *
+ * @tparam T            floating point type
+ * @tparam Dataset
+ * @param d             Dataset
+ * @param first         first locally assigned particle index of buffers in @p d
+ * @param last          first locally assigned particle index of buffers in @p d
+ * @param viewTheta     viewing angle to determine ...
+ * @param viewPhi       viewing angle to determine ...
+ * @return              array containing the observables and the second derivative of the quadpole momentum
+ */
+template<class T, class Dataset>
+std::array<T, 8> gravRad(Dataset& d,size_t first, size_t last, T viewTheta, T viewPhi)
+{
+    T dot2ibartt;
+    T dot2ibarpp;
+    T dot2ibartp;
+
+    T g = 6.6726e-8;    //Gravitational constant
+    T c = 2.997924562e10;   //Speed of light
+    T gwunits = g / std::pow(c, 4) / 3.08568025e22; //Gravitational Waves unit at 10kpc
+
+    std::array<T, 6> d2Local;
+
+    d2Local[0] = d2QuadpoleMomentum<T>(first, last, 0, 0, d); //ixx
+    d2Local[1] = d2QuadpoleMomentum<T>(first, last, 1, 1, d); //iyy
+    d2Local[2] = d2QuadpoleMomentum<T>(first, last, 2, 2, d); //izz
+
+    d2Local[3] = d2QuadpoleMomentum<T>(first, last, 0, 1, d); //ixy
+    d2Local[4] = d2QuadpoleMomentum<T>(first, last, 0, 2, d); //ixz
+    d2Local[5] = d2QuadpoleMomentum<T>(first, last, 1, 2, d); //iyz
+
+
+    int rootRank = 0;
+    std::array<T, 6> d2Global;
+    MPI_Reduce(d2Local.data(), d2Global.data(),6, MpiType<T>{}, MPI_SUM, rootRank, MPI_COMM_WORLD);
+
+    if (viewTheta == 0.0 && viewPhi == 0.0)
```

Comment:
1. Is this `if` really needed, doesn't the case below for theta, phi != 0 also work if you pass theta,phi == 0? Note how this opens a coverage gap in the test that you implemented: the case theta, phi == 0 is not tested.
2. The calculation of `httplus, httcross` from d2Global should be its own function with a unit test.